core/lib/merkle_tree/src/hasher/proofs.rs

```diff
@@ -14,26 +16,38 @@ use crate::{
 impl BlockOutputWithProofs {
     /// Verifies this output against the trusted old root hash of the tree and
     /// the applied instructions.
-    ///
-    /// # Panics
-    ///
-    /// Panics if the proof doesn't verify.
+    #[must_use = "this method returns a boolean value, use it to check the result"]
     pub fn verify_proofs(
         &self,
         hasher: &dyn HashTree,
         old_root_hash: ValueHash,
         instructions: &[TreeInstruction],
-    ) {
-        assert_eq!(instructions.len(), self.logs.len());
+    ) -> bool {
```

Comment:
Bikeshedding: This doesn't look idiomatic and creates unnecessary complications on the caller side. It'd be better to return a `Result`; then, you'd get `must_use` automatically, would be able to `unwrap()` in tests and handle errors as you'd like on the caller side (e.g., log them with `tracing::error!`). The crate generally defines domain-specific errors, but if you don't feel like defining them in this case, using `anyhow::Error` would be an adequate alternative. IMO, errors should also contain more context (e.g., the number of instructions and logs for the first error) to simplify debugging.

Reply:
Thinking about it, `anyhow::ensure!` with detailed information should be good enough... Errors here are not actionable anyway.

Reply:
changed